src/app/clusters/commodity-tariff-server/CommodityTariffAttrsDataMgmt.h

```diff
@@ -0,0 +1,942 @@
+
+    bool HasChanged()
+    {
+        if constexpr (IsValueNullable())
+        {
+            return NullableNotEqual(GetNewValueRef(), GetValueRef());
+        }
+        else if constexpr (IsValueList())
+        {
+            return ListsNotEqual(GetNewValueRef(), GetValueRef());
+        }
+        else
+        {
+            return GetNewValueRef() != GetValueRef();
+        }
+    }
+
+    void CleanupValue(ValueType & aValue)
+    {
+        if constexpr (IsValueNullable())
+        {
+            if (!aValue.IsNull())
+            {
+                if constexpr (IsList<WrappedType>::value)
+                {
+                    CleanupList(aValue.Value());
+                }
+                else if constexpr (IsStruct<WrappedType>::value)
+                {
+                    CleanupStruct(aValue.Value());
+                }
+            }
+            aValue.SetNull();
+        }
+        else if constexpr (IsValueList())
+        {
+            CleanupList(aValue);
+        }
+    }
+
+    void CleanupList(DataModel::List<PayloadType> & list)
+    {
+        for (auto & item : list)
+        {
+            CleanupStruct(item);
+        }
+        if (list.data())
+        {
+            chip::Platform::MemoryFree(list.data());
+            list = DataModel::List<PayloadType>();
+        }
+    }
+
+    void CleanupStruct(PayloadType & aValue) { CleanupStructValue(aValue); }
+
+protected:
+    T mValueStorage[2]; // Double-buffered storage
+
+    void * mAuxData = nullptr;                  // Validation context
+    const uint32_t mAttrId;                     // Attribute identifier
+    void (*mAuxCb)(uint32_t, void *) = nullptr; // Update callback
+
+    StorageState mHoldState  = StorageState::kEmpty;
+    UpdateState mUpdateState = UpdateState::kIdle;
+
+    virtual CHIP_ERROR Validate(const ValueType & aValue) const { return CHIP_NO_ERROR; }
+
+    /**
+     * @brief Compares two structured values for inequality
+     * @param source The new/input value being compared
+     * @param destination The existing/stored value being compared against
+     * @return bool
+     *   - true if values are different (needs update)
+     *   - false if values are identical (no update needed)
+     * @note This intentionally uses "Compare" in the name but implements "not equal" semantics
+     *       to match the expected behavior in change detection flows.
+     * @warning Must be overridden for any struct payload types. The base implementation
+     *          always returns false (no change) and logs an error.
+     *
+     * Example override:
+     * @code
+     * bool CompareStructValue(const MyStruct& src, const MyStruct& dst) const override {
+     *     return src.field1 != dst.field1 ||
+     *            src.field2 != dst.field2;
+     * }
+     * @endcode
+     */
+    virtual bool CompareStructValue(const PayloadType & source, const PayloadType & destination) const
```

Comment:
Should probably be called StructsNotEqual or something like that, to make it clear at callsites what the return value means.